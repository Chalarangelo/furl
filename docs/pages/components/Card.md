---
title: Card
description: furl's Card component.
---

Cards provide a versatile container for any sort of content

### Examples

<cardexamples></cardexamples>

```jsx
import React from 'react';
import ReactDOM from 'react-dom';
import { Card, CardSection, Title, Text, Paragraph, Button } from 'furl-components';

ReactDOM.render(
  <Card width={320}>
    <CardSection>
      <Title level={5}>Cool mountains<Text textStyle='small'>Just look at them</Text></Title>
    </CardSection>
    <CardSection media='mountain-scenery.jpg' height={200} />
    <CardSection>
      <Paragraph>This is a lovely image of some mountain scenery.</Paragraph>
    </CardSection>
    <Button>I want to visit the mountains</Button>
  </Card>, 
  document.getElementById('root')
);
```

### API

<table>
  <caption>Card props</caption>
  <thead>
    <tr>
      <th>Property</th>
      <th colspan="3">Description</th>
      <th>Type</th>
      <th>Default</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td class="font-c">width</td>
      <td colspan="3">card width</td>
      <td>int/string</td>
      <td class='font-c'>'auto'</td>
    </tr>
  </tbody>
</table>

<table>
  <caption>CardSection props</caption>
  <thead>
    <tr>
      <th>Property</th>
      <th colspan="3">Description</th>
      <th>Type</th>
      <th>Default</th>
    </tr>
  </thead>
  <tbody>
   <tr>
      <td class="font-c">media</td>
      <td colspan="3">card section media url</td>
      <td>string</td>
      <td class='font-c'></td>
    </tr>
    <tr>
      <td class="font-c">height</td>
      <td colspan="3">card section height</td>
      <td>int/string</td>
      <td class='font-c'>'auto'</td>
    </tr>
  </tbody>
</table>

### Notes

* A `Card` only accepts children of the `CardSection` component type. Children that are not of the `CardSection` component type will be automatically cast to appropriate components.
* The `CardSection`'s `height` accepts any valid CSS value, including CSS variables, and will only be applied if `media` is specified.
* If a `CardSection` has a `media` prop specified, it should not have any children.
* `width` and `height` accept any valid CSS value, including CSS variables.