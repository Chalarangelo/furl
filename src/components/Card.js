import React from "react";
// import "./Card.css";

const Card = props => <div className="card">{props.children}</div>;

export default Card;