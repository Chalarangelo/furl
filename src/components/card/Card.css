.card {
  border: 1px solid #ddd;
  background: white;
  color: black;
  padding: 6px 8px;
  border-radius: 4px; }
