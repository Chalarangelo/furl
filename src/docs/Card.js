import React from "react";
import Hyperlink from "../components/Hyperlink";
import {Table, TableCaption, TableHead, TableRow, TableCell} from "./components/Table";

const CardDocs = props => (
  <div>
    <h2>Card</h2>
    <p>Cards allow you to create rectangular containers for any kind of content.</p>
    <h3>Examples</h3>
    {/* TODO: Add component demo */}
    <h3>API</h3>
    <p>Cards are made up of a <code>&lt;Card&gt;</code> component, which is the card's container element and one or more <code>&lt;CardSection&gt;</code> components, which are the sections of the card, passed as <code>children</code> to the <code>&lt;Card&gt;</code> component.</p>
    <Table>
      <TableCaption>Card props</TableCaption>
      <TableHead>
        <TableRow>
          <TableCell heading>Name</TableCell>
          <TableCell heading>Type</TableCell>
          <TableCell heading>Default</TableCell>
          <TableCell heading>Description</TableCell>
        </TableRow>
      </TableHead>
      <TableBody>
        <TableRow>
          <TableCell><code>width</code></TableCell>
          <TableCell><code>string</code></TableCell>
          <TableCell><code>'auto'</code></TableCell>
          <TableCell>width of the card</TableCell>
        </TableRow>
        <TableRow>
          <TableCell><code>id</code></TableCell>
          <TableCell><code>string</code></TableCell>
          <TableCell>-</TableCell>
          <TableCell>value for the rendered element's <code>id</code></TableCell>
        </TableRow>
        <TableRow>
          <TableCell><code>className</code></TableCell>
          <TableCell><code>string</code></TableCell>
          <TableCell>-</TableCell>
          <TableCell>additional classes that should be passed to the component</TableCell>
        </TableRow>
        <TableRow>
          <TableCell><code>children</code></TableCell>
          <TableCell><code>React.children</code></TableCell>
          <TableCell>-</TableCell>
          <TableCell>the component's children elements (only accepts <code>&lt;CardSection&gt;</code> elements)</TableCell>
        </TableRow>
      </TableBody>
    </Table>
    <Table>
      <TableCaption>CardSection props</TableCaption>
      <TableHead>
        <TableRow>
          <TableCell heading>Name</TableCell>
          <TableCell heading>Type</TableCell>
          <TableCell heading>Default</TableCell>
          <TableCell heading>Description</TableCell>
        </TableRow>
      </TableHead>
      <TableBody>
        <TableRow>
          <TableCell><code>media</code></TableCell>
          <TableCell><code>string</code></TableCell>
          <TableCell>-</TableCell>
          <TableCell>url source of the image, if any</TableCell>
        </TableRow>
        <TableRow>
          <TableCell><code>height</code></TableCell>
          <TableCell><code>string</code></TableCell>
          <TableCell><code>'auto'</code></TableCell>
          <TableCell>height of the section</TableCell>
        </TableRow>
        <TableRow>
          <TableCell><code>id</code></TableCell>
          <TableCell><code>string</code></TableCell>
          <TableCell>-</TableCell>
          <TableCell>value for the rendered element's <code>id</code></TableCell>
        </TableRow>
        <TableRow>
          <TableCell><code>className</code></TableCell>
          <TableCell><code>string</code></TableCell>
          <TableCell>-</TableCell>
          <TableCell>additional classes that should be passed to the component</TableCell>
        </TableRow>
        <TableRow>
          <TableCell><code>children</code></TableCell>
          <TableCell><code>React.children</code></TableCell>
          <TableCell>-</TableCell>
          <TableCell>the component's children elements (will be ignored if <code>media</code> is set)</TableCell>
        </TableRow>
      </TableBody>
    </Table>
  </div>
);

export default CardDocs;
