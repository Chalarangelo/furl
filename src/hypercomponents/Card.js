import React from 'react';
import { Card, CardSection } from '../components';
import { hasKey, omitProps } from '../utilities/utils';

const CardHOC = (props) => {
  if (!hasKey(props, 'data'))
    return (
      <Card {...props}/>
    );

  let data = props.data;
  console.log(data);

  return (
    <Card {...omitProps(props, ['data'])}>
      {data.map((val, i) => {
        let children = hasKey(val, 'content') ? val.content : hasKey(val, 'children') ? val.children : val;
        return (
          <CardSection {...omitProps(val, ['content', 'children'])} key={`sc_${i}`} >
            {children}
          </CardSection>
        )
      })}
    </Card>
  );
};

export default CardHOC;
