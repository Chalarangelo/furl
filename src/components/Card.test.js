import React from "react";
import Card from "./Card";
import renderer from "react-test-renderer";

describe("Card", () => {
  it("renders properly", () => {
    const tree = renderer
      .create(<Card>Hello</Card>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});