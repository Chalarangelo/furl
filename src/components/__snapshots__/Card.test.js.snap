// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card renders properly 1`] = `
<div
  className="card"
>
  Hello
</div>
`;
